### ISSUE TITLE
[HOLD for payment 2023-03-01] [$1000] Editing header bold text in chat increases vertical gap above the text

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14668]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app
2. Open any chat
3. Send bold text using #
4. Edit the text

## Expected Result:
Vertical gap above the text should not change after edit

## Actual Result:
Vertical gap above the text is increased after editing

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.61-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/215596710-75e3d572-746f-4c2a-a4ed-bf90de2c6368.mp4

https://user-images.githubusercontent.com/43996225/215596739-6b89b0fa-b53d-400f-a285-258b4ba4c432.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675067869948869

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011432c27e94a6da98</li>
        <li>Upwork Job ID: 1623338262676652032</li>
        <li>Last Price Increase: 2023-02-08</li>
    </ul>
</details>

**Comment by dylanexpensify** [comment]
@shawnborton can you confirm this is indeed a bug, and I'll get it moved over?

**Comment by dylanexpensify** [comment]
Shawn is OOO, but back this week!

**Comment by shawnborton** [comment]
Hm yeah, this seems like a bug! It's a bit strange that there was even a blank line inserted above the #/headline in the first place.

**Comment by dylanexpensify** [comment]
Replicated

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~011432c27e94a6da98

**Comment by MelvinBot** [comment]
Current assignee @dylanexpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @eVoloshchak (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @danieldoglas (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by tienifr** [comment]
## Proposal

### Problem
After editing a bold text, we can see vertical gap above the text is increased. The reason is an unexpected line break was added before the edited text. 
### RCA
This issue happens because when we handle edited text, https://github.com/Expensify/App/blob/a171133dbbda4552366511fc2acedec1816e6137/src/libs/actions/Report.js#L821-L832  we convert HTML to markdown and `parser.htmlToMarkdown(htmlWithAutoLinks)` adds two redundant end-of-line `\n` before and after our markdown (check the logic [here](https://github.com/Expensify/expensify-common/blob/51cfcf97db590bc416d858a762ddd7c72d39841d/lib/ExpensiMark.js#L231-L233)).https://github.com/Expensify/App/blob/f8d7a3a8e291cd216b26b22e9e8a69c93bb15e9e/src/libs/actions/Report.js#L806-L809 

But why there is only a line break before the edited text? Normally, the code for end-of-line `\n` will be replaced by `<br />` when we convert the markdown back to HTML so we should see a line break before and a line break after the edited text. While investigating I found that the second code for end-of-line `\n` is removed by chance when we convert the markdown to HTML. https://github.com/Expensify/App/blob/a171133dbbda4552366511fc2acedec1816e6137/src/libs/actions/Report.js#L830-L831

### Solution 
We need to remove two redundant `\n` before we convert markdown to HTML. After that, there will be no unexpected line breaks . We will trim `markdownForNewComment` as below:
```diff
diff --git a/src/libs/actions/Report.js b/src/libs/actions/Report.js
index 7abb23497..2172559dc 100644
--- a/src/libs/actions/Report.js
+++ b/src/libs/actions/Report.js
@@ -825,8 +825,8 @@ function editReportComment(reportID, originalReportAction, textForNewComment) {
     // https://github.com/Expensify/App/issues/9090
     // https://github.com/Expensify/App/issues/13221
     const originalCommentHTML = lodashGet(originalReportAction, 'message[0].html');
-    const markdownForNewComment = handleUserDeletedLinks(textForNewComment, originalCommentHTML);
-
+    let markdownForNewComment = handleUserDeletedLinks(textForNewComment, originalCommentHTML);
+    markdownForNewComment = markdownForNewComment.trim();
     const autolinkFilter = {filterRules: _.filter(_.pluck(parser.rules, 'name'), name => name !== 'autolink')};
     const htmlForNewComment = parser.replace(markdownForNewComment, autolinkFilter);
     const parsedOriginalCommentHTML = parser.replace(parser.htmlToMarkdown(originalCommentHTML), autolinkFilter);
```

### Result 

https://user-images.githubusercontent.com/113963320/216362743-1db9f146-12ff-43ae-96c1-36e4ccbff59e.mov

**Comment by hackykitty** [comment]
I reviewed @eh2077 and @Tienifr's proposals. Great.
But elsewhere the string edited is laid horizontally, but adding bold puts it vertically.
To fix this, you need to follow the style:
## flex-direction: row;
## align-items: center;
Applying style eliminates the need to remove two duplicate \n's before converting the markdown to HTML.
All problems can be solved by adding only 2 styles.
thanks.

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Vertical gap above the text is increased after editing a `h1` tag

### What is the root cause of that problem?
In `htmlToMarkdown`, we're adding new line to both before and after the `# `, which should not be the case, we should only add new lines if there's non-new-line content before and after the heading.

Why non-new-line? because if before the heading there's already a new line, the `# ` will be on the next line already and we should not add a (redundant) new line in this case. This complies with how HTML renders the `h1` tag.

### What changes do you think we should make in order to solve the problem?


In [here](https://github.com/Expensify/expensify-common/blob/668be7533215fc8d683d658da839361664320965/lib/ExpensiMark.js#L232), we should use a regex that satisfies the requirement `we should only add new lines if there's non-new-line content before the heading`.

```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index bac4b81..5622980 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -229,8 +229,8 @@ export default class ExpensiMark {
             },
             {
                 name: 'heading1',
-                regex: /\s*<(h1)(?:"[^"]*"|'[^']*'|[^'">])*>(.*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))\s*/gi,
-                replacement: '\n# $2\n',
+                regex: /([^\r\n]*?)\s*<(h1)(?:"[^"]*"|'[^']*'|[^'">])*>(.*?)<\/\2>(?![^<]*(<\/pre>|<\/code>))\s*([^\r\n]*)/gi,
+                replacement: (match, g1, g2, g3, g4, g5) => `${g1}${g1 && '\n'}# ${g3}${g5 && '\n'}${g5}`,
             },
             {
                 name: 'listItem',

```

After this fix, another issue will occur. Let's consider this HTML
`<h1>heading 1</h1><h1>heading 2</h1>`. In here if both `h1` blocks matches the regex, it will add 2 new lines in between the blocks, one after `heading 1` and one before `heading 2`. This is because in the match, `heading 1` has non-new-line content after it and `heading 2` has non-new-line content before it. This is incorrect understanding since after `heading 1` is matched, it will add 1 new line after it and `heading 2` no longer has non-new-line content before it. 

In order to fix this we need to introduce another property into the `rule`, which can be `sequential: true`. If the rule is sequential, the regex replacement will be applied once at a time. Example of this in `htmlToMarkdown`:

```
if (rule.sequential) {
                while (rule.regex.test(generatedMarkdown)) {
                    generatedMarkdown = generatedMarkdown.replace(rule.regex, rule.replacement);
                }
            } else {
                generatedMarkdown = generatedMarkdown.replace(rule.regex, rule.replacement);
            }
```

Of course the regex above will need to have the global flag `/g` removed for this to work.

### What alternative solutions did you explore? (Optional)

**Comment by eh2077** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/14668#issuecomment-1422748299)

To fix the following test case mentioned by @tienifr 's in his latest proposal https://github.com/Expensify/App/issues/14668#issuecomment-1423712701
```html
<h1>heading 1</h1><h1>heading 2</h1>
```

I'd like to update the solution section of my previous proposal
### What changes do you think we should make in order to solve the problem?

I'd like to propose a new solution to improve the regex rule which translates html `<h1>` header to markdown. The new solution is to replace 

https://github.com/Expensify/expensify-common/blob/668be7533215fc8d683d658da839361664320965/lib/ExpensiMark.js#L233

with 
```javascript
replacement: '[block]# $2[block]',
```
By replacing 
```
\n# $2\n
``` 
with 
```
[block]# $2[block]
```
we can reuse this function `ExpensiMark#replaceBlockWithNewLine` to handle line break properly. See

https://github.com/Expensify/expensify-common/blob/668be7533215fc8d683d658da839361664320965/lib/ExpensiMark.js#L521

https://github.com/Expensify/expensify-common/blob/668be7533215fc8d683d658da839361664320965/lib/ExpensiMark.js#L455-L467

>    /**
     * replace [block] with '\n' if :
     * 1. We have text before the [block].
     * 2. The text does not end with a new line.
     * 3. The text does not have quote mark '>' .
     * 4. [block] not the last [block] in the string.
     *
     * @param {String} htmlString
     * @returns {String}
     */
    replaceBlockWithNewLine(htmlString) {

**Comment by danieldoglas** [comment]
@eVoloshchak what's your saying on these proposals?

**Comment by eVoloshchak** [comment]
I see a couple of good proposals, reviewing tomorrow

**Comment by danieldoglas** [comment]
Thanks @eVoloshchak

**Comment by MelvinBot** [comment]
@danieldoglas @eVoloshchak @dylanexpensify this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by eVoloshchak** [comment]
I like @eh2077's [proposal](https://github.com/Expensify/App/issues/14668#issuecomment-1422748299), great explanation of the root cause. Both @tienifr's and @eh2077's proposals have a lot in common, but @eh2077 was the first to propose this approach. 

We'll need to test more thoroughly during the review (as I couldn't test ALL of the cases), but from my initial testing, this works as expected. 
🎀👀🎀 C+ reviewed!
cc: @danieldoglas

**Comment by eh2077** [comment]
Hi @eVoloshchak , thanks for reviewing the proposal.

I think my [updated solution](https://github.com/Expensify/App/issues/14668#issuecomment-1423883167) will be a better one to address this issue. My original solution can only handle text with only one `h1` tag. Can you also have a review on it?

**Comment by eVoloshchak** [comment]
> I think my [updated solution](https://github.com/Expensify/App/issues/14668#issuecomment-1423883167) will be a better one to address this issue. My original solution can only handle text with only one `h1` tag. Can you also have a review on it?

Yeah, I tested your updated proposal when reviewing. 
I linked the original comment because it has an explanation of the problem, but you're correct, I should include the link to your updated proposal also, thank you.

**Comment by MelvinBot** [comment]
📣 @eh2077 You have been assigned to this job by @danieldoglas!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by eh2077** [comment]
Thanks @danieldoglas!

I applied the job and PR https://github.com/Expensify/expensify-common/pull/502 for `expensify-common` is ready for review. 

cc @eVoloshchak @danieldoglas

**Comment by danieldoglas** [comment]
@eVoloshchak FYI if you comment in the original PR, I can assign you to review there directly.

**Comment by danieldoglas** [comment]
hey @eh2077 can you create a new PR updating the expensify-commons library in the app and do the full checklist on App? 
The merge commit hash is `bf099f71a09bd13cdaef24b30ee63b4d801a686c`

**Comment by eh2077** [comment]
Hi @eVoloshchak and @danieldoglas 

[PR](https://github.com/Expensify/App/pull/15241) is ready for review, thanks!

**Comment by eh2077** [comment]
Kindly bump @danieldoglas @eVoloshchak

**Comment by eVoloshchak** [comment]
Testing in a couple of hours

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.75-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15241


If no regressions arise, payment will be issued on 2023-03-01. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
📣 @MelvinBot! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@eVoloshchak / @danieldoglas] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@eVoloshchak / @danieldoglas] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@eVoloshchak / @danieldoglas] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@eh2077] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@dylanexpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by eh2077** [comment]
### Regression Test Proposal

1. Open a chat and send this message

   ```text
   # heading
   ```

2. Edit the sent message and verify it's displayed in one line in editing mode

3. Edit and change the sent message to

   ```text
   # heading test
   ```
   and save it

4. Verify the edited header bold text is still displayed in one line

Do we agree 👍 or 👎

**Comment by MelvinBot** [comment]
📣 @eh2077! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by eh2077** [comment]
Contributor details
Your Expensify account email: hanyl_fight@126.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0184a9d594d4a0e25a

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by danieldoglas** [comment]
@eVoloshchak can you fill this https://github.com/Expensify/App/issues/14668#issuecomment-1440926211? Thanks!

**Comment by danieldoglas** [comment]
@dylanexpensify I think we should include this as part of the tests related to the markdown edits.

**Comment by eVoloshchak** [comment]
> @eVoloshchak can you fill this [#14668 (comment)](https://github.com/Expensify/App/issues/14668#issuecomment-1440926211)? Thanks!

Sure, doing it today

**Comment by eVoloshchak** [comment]
- The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/expensify-common/pull/468

- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: the PR description mentioned the [design doc](https://docs.google.com/document/d/1YJmMi9NjnstCAP2Ibo_IZdT2T16Res2jdhWr2WSrk50), which I can't access, I can't verify if it's a regression or was by design at the time

- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: automated regression tests will catch this

**Comment by dylanexpensify** [comment]
Whipping up RT and payments today

**Comment by danieldoglas** [comment]
@dylanexpensify I think we're being very generalistic in `Edit a message with Markdown`. the test suggested here will test exactly for the situation where editing a header, so I think it makes sense to update it.

**Comment by dylanexpensify** [comment]
Bump @danieldoglas! 🙇‍♂️

**Comment by eh2077** [comment]
> @dylanexpensify I think we're being very generalistic in `Edit a message with Markdown`. the test suggested here will test exactly for the situation where editing a header, so I think it makes sense to update it.

Hi @dylanexpensify , is this comment from @danieldoglas you're waiting for?

**Comment by dylanexpensify** [comment]
Wow, yes @eh2077 haha, I didn't refresh issue before posting 🤦

**Comment by dylanexpensify** [comment]
RT linked above! Working on payments now!

**Comment by dylanexpensify** [comment]
@eh2077 sent offer!
@eVoloshchak sent invite to apply to job!
@dhanashree-sawant sent invite to apply to job as reporter!

**Comment by dylanexpensify** [comment]
@eh2077 payment sent!
@eVoloshchak offer sent!

**Comment by dhanashree-sawant** [comment]
@dylanexpensify Thanks and accepted the offer!

**Comment by dylanexpensify** [comment]
@dhanashree-sawant please apply when you can!

**Comment by dylanexpensify** [comment]
Offer sent @dhanashree-sawant !

**Comment by dylanexpensify** [comment]
alrightyyyy we're all settled up!! Thanks everyone!!
